game/hud/src/components/Compass/CompassPOIManager.tsx

```diff
@@ -0,0 +1,462 @@
+                  [poi.id]: poi,
+                };
+              }
+            }, {})),
+          },
+        },
+      };
+    });
+  }
+
+  private removePOIByType = (type: string) => {
+    this.setState((prevState: CompassPOIManagerInnerState) => {
+      return {
+        ...prevState,
+        poiList: {
+          ...(Object.keys(prevState.poiList).reduce((carry, key) => {
+            if (key === type) {
+              return {
+                ...carry,
+                [key]: {},
+              };
+            } else {
+              return {
+                ...carry,
+                [key]: prevState.poiList[key],
+              };
+            }
+          }, {})),
+        },
+      };
+    });
+  }
+
+  private getBearing = (target: Vec2f | Vec3f | number) => {
+    if (typeof target === 'number') {
+      return 0;
+    } else {
+      return calculateBearing(this.props.position, target);
+    }
+  }
+
+  private getDistance = (target: Vec2f | Vec3f | number) => {
+    if (typeof target === 'number') {
+      return 0;
+    } else {
+      return calculateDistance(this.props.position, target);
+    }
+  }
+
+  private getElevation = (target: Vec2f | Vec3f | number) => {
+    if (typeof target === 'number') {
+      return 0;
+    } else {
+      return calculateElevation(this.props.position, target);
+    }
+  }
+
+  private isTargetAbove = (target: Vec2f | Vec3f | number, byAtLeast: number = 0) => {
+    return (this.getElevation(target) - byAtLeast) > 0;
+  }
+
+  private isTargetBelow = (target: Vec2f | Vec3f | number, byAtLeast: number = 0) => {
+    return (this.getElevation(target) + byAtLeast)  < 0;
+  }
+
+  private isTargetLevel = (target: Vec2f | Vec3f | number, buffer: number = 0) => {
+    return this.getElevation(target) <= buffer &&  this.getElevation(target) >= (0 - buffer);
+  }
+
+  private getPoiPlacement = (angleOrPosition: number | Vec3f | Vec2f): number => {
+    let angle;
+    if (typeof angleOrPosition === 'number') {
+      angle = angleOrPosition;
+    } else {
+      angle = this.getBearing(angleOrPosition);
+    }
+    angle = angle % 360;
+    const facingAdjustment = 360 - this.state.facingNorth;
+    const percentPerDegree = 100 / this.props.degreesToShow;
+    const adjustedAngle = convertToMinusAngle((angle + facingAdjustment) % 360);
+    const leftPosition = (adjustedAngle * percentPerDegree) + 50;
+    return leftPosition;
+  }
+
+  private getPoiPlacementStyle = (placement: number, offset: number = 0): React.CSSProperties => {
+    return {
+      visibility: (placement < -25 || placement > 125) ? 'hidden' : 'visible',
+      left: `calc(${placement}% - ${offset}px)`,
+    };
+  }
+}
+
+export interface CompassPOIManagerProps {
+  children: (value: CompassContext) => React.ReactNode;
+  degreesToShow: number;
+}
+
+export interface CompassPOIManagerState {
+  position: Vec3f;
+}
+
+class CompassPOIManager extends React.Component<CompassPOIManagerProps, CompassPOIManagerState> {
```

Comment:
implement shouldComponentUpdate here or make it a PureComponent

Reply:
added `shouldComponentUpdate` for this

Reply:
Thanks